### PR TITLE
DCE inference graphs too

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97275

I added a bunch of asserts to verify that I didn't accidentally kill copy_ in the graph, hopefully this combined with our existing tests is good enough.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

File: torch/_functorch/aot_autograd.py [hunk 1]
```diff
@@ -1283,11 +1287,27 @@ def aot_dispatch_base(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig, *
...
 
 
-def assert_functional_graph(fx_g: torch.fx.Graph):
+# Returns the number of detected copy_
+def assert_functional_graph(fx_g: torch.fx.Graph, *, allow_input_mutations: bool = False) -> int:
+    placeholders = set()
+    copy_count = 0
+    # NB: It would also be nice to verify that the mutations all happen at the
+    # end, but we also do some administrative views after mutations so this
+    # isn't actually true.  (TODO: Could this cause problems for Inductor?)
```
Comment (bdhirsh) [inline torch/_functorch/aot_autograd.py:1296]: Hmm, I vaguely remember this being a problem, but I thought we "fixed" it by syncing both the inputs and outputs before doing any copies during tracing. But I might be wrong (do you know of an instance where this happens?).

File: torch/_functorch/aot_autograd.py [hunk 2]
```diff
@@ -1258,12 +1258,16 @@ def aot_dispatch_base(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig, *
...
+    # As long as we opted to remove input mutations, then
+    # there should be *NO* mutating ops in the graph at this point.
+    copy_count = assert_functional_graph(fw_module.graph, allow_input_mutations=aot_config.keep_inference_input_mutations)
+
+    fw_module.graph.eliminate_dead_code()
+    fw_module.recompile()
+
+    copy_count2 = assert_functional_graph(fw_module.graph, allow_input_mutations=aot_config.keep_inference_input_mutations)
+
+    assert copy_count == copy_count2
```
Comment (bdhirsh) [inline torch/_functorch/aot_autograd.py:1270]: seems very reasonable

**Review by bdhirsh — APPROVED** [review]
lgtm!

File: torch/_functorch/aot_autograd.py (hunk 1 above)
Comment (ezyang) [reply torch/_functorch/aot_autograd.py:1296]: yeah it failed on `pytest test/functorch/test_aotdispatch.py -k test_input_output_view_mutate_multiple`

**Pull request closed without merging**